### PR TITLE
Add SPEED-bench

### PR DESCRIPTION
**Pull request by talorabr**
PR to support speculative decoding (SD) evaluation, and specifically to support SPEED-Bench dataset (a new standard dataset for speculative decoding): 

* Add SD metrics.
* Add SD evaluator and generation flows.
* Add SPEED-Bench dataset for SD evaluation.

## Summary by CodeRabbit

* **New Features**
  * End-to-end Speculative Decoding (SD) evaluation: metric collection, delta computation, generation-time injection, and evaluator integration
  * SPEED-Bench data preparation CLI to generate per-config evaluation datasets
  * New SD reporting metrics: acceptance length, acceptance rate, and per-position statistics

* **Documentation**
  * Added Speculative Decoding evaluation guide with examples and benchmark setup

* **Chores**
  * Improved shell-argument quoting in server launch scripts

### CODE REVIEW

File: docs/evaluation/speculative-decoding.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Speculative Decoding
```
Comment (Kipok) [inline docs/evaluation/speculative-decoding.md:1]: Please add this file to mkdocs.yml so that it's visible in the navigation on the website

File: docs/evaluation/speculative-decoding.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+ns eval \
+    --cluster=<cluster config> \
+    --data_dir=<must match prepare_data parameter> \
+    --output_dir=<any mounted output location> \
+    --benchmarks=speed-bench \
+    --model=meta-llama/Llama-3.3-70B-Instruct \
+    --server_args="--speculative-algorithm STANDALONE --speculative-draft-model-path meta-llama/Llama-3.2-1B-Instruct --speculative-num-steps 3 --speculative-eagle-topk 1 --torch-compile-max-bs 32 --max-running-requests 32 --cuda-graph-max-bs 32 --mem-fraction-static 0.8" \
+    --server_nodes=1 \
+    --server_gpus=8 \
+    --server_type=sglang
```
Comment (Kipok) [inline docs/evaluation/speculative-decoding.md]:
```suggestion
    --server_type=sglang \
```

File: docs/evaluation/speculative-decoding.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ns eval \
+    --cluster=<cluster config> \
+    --data_dir=<must match prepare_data parameter> \
+    --output_dir=<any mounted output location> \
+    --benchmarks=speed-bench \
+    --model=meta-llama/Llama-3.3-70B-Instruct \
+    --server_args="--speculative-config '{\"method\": \"eagle3\", \"num_speculative_tokens\": 3, \"model\": \"nvidia/Llama-3.3-70B-Instruct-Eagle3\"}'" \
+    --server_nodes=1 \
+    --server_gpus=8 \
+    --server_type=vllm
```
Comment (Kipok) [inline docs/evaluation/speculative-decoding.md]:
```suggestion
    --server_type=vllm \
```

File: nemo_skills/dataset/speed-bench/__init__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# settings that define how evaluation should be done by default (all can be changed from cmdline)
+REQUIRES_DATA_DIR = True
+DATASET_GROUP = "speculative-decoding"
```
Comment (Kipok) [inline nemo_skills/dataset/speed-bench/__init__.py]:
```suggestion
```
this has been deprecated recently

File: nemo_skills/dataset/speed-bench/prepare.py [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+        if config_name == "train_test_split":
+            EXTERNAL_DATASETS[full_name] = (
+                EXTERNAL_DATASETS[full_name]["train"],
+                EXTERNAL_DATASETS[full_name]["test"],
+            )
+    return EXTERNAL_DATASETS[full_name]
+
+
+@staticmethod
+def _generate_stackselect_prompt(question: str, answers: list[str], answer: str, num_tokens: int) -> str:
```
[the remark this answers is not in the document]
Comment (Kipok) [reply nemo_skills/dataset/speed-bench/prepare.py]: I guess that's probably a good idea to do, @talorabr

File: nemo_skills/evaluation/evaluator/specdec.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            stats.get("acceptance_length", 0.0),
+            stats.get("acceptance_rate", 0.0),
+            stats.get("num_drafts", 0),
+        )
+        for sample in data:
+            for key, value in stats.items():
+                if key not in sample:
+                    sample[key] = value
+    else:
+        LOG.warning(
```
Comment (Kipok) [inline nemo_skills/evaluation/evaluator/specdec.py]: should we fail instead? If this is a spec decoding benchmark, I think it's better to just fail with that error

File: nemo_skills/evaluation/evaluator/specdec.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    # ------------------------------------------------------------------
+    # 2. Inject pre-computed spec-decode stats into each data point
+    # ------------------------------------------------------------------
+    stats = eval_config.specdec_stats
+    if stats is not None:
+        LOG.info(
+            "Stamping spec-decode stats onto %d data points: "
+            "acceptance_length=%.4f, acceptance_rate=%.2f%%, num_drafts=%d",
+            len(data),
+            stats.get("acceptance_length", 0.0),
```
Comment (Kipok) [inline nemo_skills/evaluation/evaluator/specdec.py]: if those are always expected to be present, then better to access as `stats["acceptance_length"]` and fail if it's not there

File: nemo_skills/evaluation/metrics/specdec_metrics.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      computed as ``delta_per_pos[i] / delta_drafts``.
+
+    """
+
+    def __init__(self):
+        super().__init__(compute_no_answer=False)
+
+    def _get_score_dict(self, prediction: dict) -> dict[str, bool | int | float]:
+        return {
+            "spec_draft_tokens": prediction.get("draft_tokens", 0),
```
Comment (Kipok) [inline nemo_skills/evaluation/metrics/specdec_metrics.py]: if those are expected to always be present, better to access as `prediction["draft_tokens"]` and fail if they aren't present

File: nemo_skills/inference/eval/specdec.py [hunk 9]
```diff
@@ -0,0 +1,784 @@
...
+                            )
+                            / len(data_point["response_ids"]),
+                            "acceptance_length": sum(
+                                metrics[response_id]["acceptance_length"] for response_id in data_point["response_ids"]
+                            )
+                            / len(data_point["response_ids"]),
+                        }
+                    )
+                else:
+                    LOG.warning(
```
Comment (Kipok) [inline nemo_skills/inference/eval/specdec.py:580]: should we fail here instead?

File: nemo_skills/evaluation/evaluator/specdec.py (hunk 6 above)
Comment (talorabr) [reply nemo_skills/evaluation/evaluator/specdec.py]: Changed in the configuration to check if `specdec_stats` is a dictionary and raises an error if not

File: nemo_skills/inference/eval/specdec.py (hunk 9 above)
Comment (talorabr) [reply nemo_skills/inference/eval/specdec.py:580]: Not necessarily. There was a bug in SGLang where concurrent writes to the metrics file could corrupt it. This has been fixed in: https://github.com/sgl-project/sglang/pull/19011 .
In addition, we have a fallback mechanism that retrieves metrics from the `/metrics` endpoint if reading from the file fails.
The benefit of using the metrics file is that it provides more detailed metrics compared to the `/metrics` endpoint.

**Comment by gwarmstrong** [comment]
closing in favor of  #1279
